### PR TITLE
Be able to upload attachment to Location

### PR DESCRIPTION
**Pull request by mk-simsoft**
ANET now supports uploading attachments to existing Locations.

Closes [AB#877](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/877), [AB#882](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/882), [AB#883](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/883)

#### User changes
- None
#### Superuser changes
- Superuser can upload attachments to locations.
#### Admin changes
- Admin can upload attachments to locations.

#### System admin changes
- [x] anet.yml or anet-dictionary.yml needs change
- [x] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [ ] Added and/or updated unit tests
  - [x] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

**Review by midmarch — CHANGES_REQUESTED** [review]
When I try to upload a picture, I am getting the following error:

image

I suggest

1. Handle error, don't leave the user in the dark, use a toaster error message when upload fails
2. I think the error is related to [AB#897](https://dev.azure.com/ncia-anet/ANET/_workitems/edit/897).

**Comment by mk-simsoft** [comment]
> When I try to upload a picture, I am getting the following error:
> 
> image
> 
> I suggest
> 
> 1. Handle error, don't leave the user in the dark, use a toaster error message when upload fails
> 2. I think the error is related to [AB#897](https://dev.azure.com/ncia-anet/ANET/_workitems/edit/897).

On this branch you should not take any error when you try uploading. [AB#897](https://dev.azure.com/ncia-anet/ANET/_workitems/edit/897) this code change can effect but not in this branch.

File: client/src/components/Attachment/UploadAttachment.js [hunk 1]
```diff
@@ -25,7 +25,12 @@ const GQL_UPDATE_ATTACHMENT = gql`
   }
 `
 
-const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
+const UploadAttachment = ({
+  getRelatedObject,
+  edit,
+  saveRelatedObject,
+  childFunctionTrigger
```
Comment (gjvoosten) [inline client/src/components/Attachment/UploadAttachment.js:32]: Is this parameter used anywhere?

File: client/src/components/Attachment/UploadAttachment.js [hunk 2]
```diff
@@ -105,10 +110,15 @@ const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
   const handleFileEvent = async e => {
     // Control related object has an uuid or not
     if (!relatedObject.uuid) {
-      saveRelatedObject().then(async response => {
-        relatedObject.uuid = response
+      if (relatedObject.type !== "reports") {
```
Comment (gjvoosten) [inline client/src/components/Attachment/UploadAttachment.js:113]: This again makes this specific to reports, which you shouldn't (have to) do.

File: client/src/components/Attachment/UploadAttachment.js [hunk 3]
```diff
@@ -105,10 +110,15 @@ const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
...
+        relatedObject.uuid = saveRelatedObject()
```
Comment (gjvoosten) [inline client/src/components/Attachment/UploadAttachment.js:114]: You should wait until the async `saveRelatedObject()` function has completed before saving the attachment.

File: client/src/components/Attachment/UploadAttachment.js [hunk 4]
```diff
@@ -159,12 +169,14 @@ const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
   function save(values, relatedObjectUuid, edit) {
     const attachment = Attachment.filterClientSideFields(values)
     const operation = edit ? GQL_UPDATE_ATTACHMENT : GQL_CREATE_ATTACHMENT
+    attachment.attachmentRelatedObjects[0].relatedObjectUuid = relatedObjectUuid
```
Comment (gjvoosten) [inline client/src/components/Attachment/UploadAttachment.js:172]: Is this necessary?

File: client/src/pages/locations/Form.js [hunk 5]
```diff
@@ -350,6 +375,10 @@ const LocationForm = ({ edit, title, initialValues, notesComponent }) => {
     navigate(-1)
   }
 
+  function saveCallback(values) {
+    return values
+  }
```
Comment (gjvoosten) [inline client/src/pages/locations/Form.js:380]: This doesn't actually save anything…

File: client/src/pages/locations/Form.js [hunk 6]
```diff
@@ -367,20 +396,24 @@ const LocationForm = ({ edit, title, initialValues, notesComponent }) => {
...
     // reset the form to latest values
     // to avoid unsaved changes prompt if it somehow becomes dirty
     form.resetForm({ values, isSubmitting: true })
-    if (!edit) {
-      navigate(Location.pathForEdit(location), { replace: true })
-    }
-    navigate(Location.pathFor(location), {
-      state: { success: "Location saved" }
-    })
+    setTimeout(() => {
```
Comment (gjvoosten) [inline client/src/pages/locations/Form.js:403]: `setTimeout()`, really?!

File: client/src/pages/locations/Show.js [hunk 7]
```diff
@@ -31,10 +31,19 @@ import { connect } from "react-redux"
 import { useLocation, useParams } from "react-router-dom"
 import Settings from "settings"
 import utils from "utils"
+import AttachmentCard from "../../components/Attachment/AttachmentCard"
```
Comment (gjvoosten) [inline client/src/pages/locations/Show.js:34]: Just import from "components/Attachment/AttachmentCard".

File: client/src/pages/locations/Show.js [hunk 8]
```diff
@@ -174,6 +183,30 @@ const LocationShow = ({ pageDispatchers }) => {
...
+                        <AttachmentCard
+                          key={index}
+                          attachment={attachment}
+                          index={index}
+                          edit={false}
+                        />
+                      ))}
+                    </div>
+                  }
+                />
```
Comment (gjvoosten) [inline client/src/pages/locations/Show.js:209]: This is the same code as used in `pages/reports/Show/js` so it is a candidate for extraction into a separate component.

File: client/tests/webdriver/baseSpecs/showLocation.spec.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      expect(await await ShowLocation.getFileData()).to.be.equal(
+        "attachL…\n-1 bytes"
+      )
+    })
+    it("We can go to the show page of Attachment", async() => {
+      if (await (await ShowLocation.getImageClick()).isClickable()) {
+        await (await ShowLocation.getImageClick()).click()
+        const currentURL = await browser.getUrl()
+        expect(await currentURL).to.be.equal(
+          "http://localhost:8180/attachments/f7cd5b02-ef73-4ee8-814b-c5a7a916685d"
```
Comment (gjvoosten) [inline client/tests/webdriver/baseSpecs/showLocation.spec.js:30]: Don't hard-code the base of the URL, it may be set to something other than `http://localhost:8180`.
I see that it is also (wrongly) used in `tests/webdriver/baseSpecs/showReport.spec.js` as well, so it should be updated there too.

File: client/src/pages/locations/Form.js [hunk 10]
```diff
@@ -60,6 +61,8 @@ const LocationForm = ({ edit, title, initialValues, notesComponent }) => {
   const { currentUser } = useContext(AppContext)
   const navigate = useNavigate()
   const [error, setError] = useState(null)
+  const [childFunctionTrigger, setChildFunctionTrigger] = useState(false)
+  const [relatedObjectUuid, setRelatedObjectUuid] = useState()
```
Comment (gjvoosten) [inline client/src/pages/locations/Form.js:65]: Do you really need a `relatedObjectUuid` state field to get the location's uuid?

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/pages/locations/Form.js (hunk 6 above)
Comment (mk-simsoft) [reply client/src/pages/locations/Form.js:403]: Yes, since we can't autoSave while working on the location attachment, I realized that I could bind the attachment to the location relatedly while the page was navigating, and I proceeded that way and thought it was a useful solution since it is not the same in other forms as in the autoSave report.

File: client/src/pages/locations/Show.js (hunk 8 above)
Comment (mk-simsoft) [reply client/src/pages/locations/Show.js:209]: Let's not split it into a new component for now, but we can evaluate it when the location is finished in the attachment, but I don't think we need to separate this part as a component again. Because although it is not that big, many similar structures exist in forms, so my opinion is not compulsory for now.

File: client/tests/webdriver/baseSpecs/showLocation.spec.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      )
+    })
+    it("We can go to the show page of Attachment", async() => {
+      if (await (await ShowLocation.getImageClick()).isClickable()) {
+        await (await ShowLocation.getImageClick()).click()
+        const currentURL = await browser.getUrl()
+        const baseUrl = currentURL.split("/").slice(0, 3).join("/")
+        expect(await currentURL).to.be.equal(
+          `${baseUrl}/attachments/f7cd5b02-ef73-4ee8-814b-c5a7a916685d`
+        )
```
Comment (gjvoosten) [inline client/tests/webdriver/baseSpecs/showLocation.spec.js:32]: I would keep it simple and just do:
```javascript
diff --git a/client/tests/webdriver/baseSpecs/showLocation.spec.js b/client/tests/webdriver/baseSpecs/showLocation.spec.js
index 77f6380e0..e13dfaf18 100644
--- a/client/tests/webdriver/baseSpecs/showLocation.spec.js
+++ b/client/tests/webdriver/baseSpecs/showLocation.spec.js
@@ -25,10 +25,8 @@ describe("Show location page", () => {
     it("We can go to the show page of Attachment", async() => {
       if (await (await ShowLocation.getImageClick()).isClickable()) {
         await (await ShowLocation.getImageClick()).click()
-        const currentURL = await browser.getUrl()
-        const baseUrl = currentURL.split("/").slice(0, 3).join("/")
-        expect(await currentURL).to.be.equal(
-          `${baseUrl}/attachments/f7cd5b02-ef73-4ee8-814b-c5a7a916685d`
+        await expect(browser).toHaveUrlContaining(
+          "/attachments/f7cd5b02-ef73-4ee8-814b-c5a7a916685d"
         )
       }
     })
```

File: client/tests/webdriver/baseSpecs/showReport.spec.js [hunk 12]
```diff
@@ -49,8 +49,9 @@ describe("Show report page", () => {
       if (await (await ShowReport.getImageClick()).isClickable()) {
         await (await ShowReport.getImageClick()).click()
         const currentURL = await browser.getUrl()
+        const baseUrl = currentURL.split("/").slice(0, 3).join("/")
         expect(await currentURL).to.be.equal(
-          "http://localhost:8180/attachments/f076406f-1a9b-4fc9-8ab2-cd2a138ec26d"
+          `${baseUrl}/attachments/f076406f-1a9b-4fc9-8ab2-cd2a138ec26d`
         )
```
Comment (gjvoosten) [inline client/tests/webdriver/baseSpecs/showReport.spec.js:55]: I would keep it simple and just do:
```javascript
diff --git a/client/tests/webdriver/baseSpecs/showReport.spec.js b/client/tests/webdriver/baseSpecs/showReport.spec.js
index ed43a82db..5dafed1bc 100644
--- a/client/tests/webdriver/baseSpecs/showReport.spec.js
+++ b/client/tests/webdriver/baseSpecs/showReport.spec.js
@@ -48,10 +48,8 @@ describe("Show report page", () => {
     it("We can go to the show page of Attachment", async() => {
       if (await (await ShowReport.getImageClick()).isClickable()) {
         await (await ShowReport.getImageClick()).click()
-        const currentURL = await browser.getUrl()
-        const baseUrl = currentURL.split("/").slice(0, 3).join("/")
-        expect(await currentURL).to.be.equal(
-          `${baseUrl}/attachments/f076406f-1a9b-4fc9-8ab2-cd2a138ec26d`
+        await expect(browser).toHaveUrlContaining(
+          "/attachments/f076406f-1a9b-4fc9-8ab2-cd2a138ec26d"
         )
       }
     })
```

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/Attachment/UploadAttachment.js [hunk 13]
```diff
@@ -88,27 +106,39 @@ const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
...
-    } else {
-      await attachmentSave(e)
+      } catch (error) {
+        toast.error(`Attachment content upload failed: ${error.message}`, {
+          autoClose: false,
+          closeOnClick: true
+        })
+      }
     }
+    await attachmentSave(e)
```
Comment (midmarch) [inline client/src/components/Attachment/UploadAttachment.js:141]: Error not handled:

image

File: client/src/pages/locations/Form.js [hunk 14]
```diff
@@ -350,6 +377,10 @@ const LocationForm = ({ edit, title, initialValues, notesComponent }) => {
     navigate(-1)
   }
 
+  function takeRelatedUuid(values) {
+    return values
+  }
+
```
Comment (midmarch) [inline client/src/pages/locations/Form.js:383]: Please have look at this.

File: client/src/components/Attachment/UploadAttachment.js [hunk 15]
```diff
@@ -88,27 +106,39 @@ const UploadAttachment = ({ getRelatedObject, edit, saveRelatedObject }) => {
...
       )
   }
 
   const handleFileEvent = async e => {
     // Control related object has an uuid or not
     if (!relatedObject.uuid) {
-      saveRelatedObject().then(async response => {
+      try {
+        const response = await saveRelatedObject()
         relatedObject.uuid = response
```
Comment (midmarch) [inline client/src/components/Attachment/UploadAttachment.js:133]:
```suggestion
        relatedObject.uuid = await saveRelatedObject()
```

File: client/src/components/Attachment/UploadAttachment.js [hunk 16]
```diff
@@ -25,11 +25,29 @@ const GQL_UPDATE_ATTACHMENT = gql`
...
   const [remove, setRemove] = useState(false)
   const [uploadedList, setUploadedList] = useState([])
   const relatedObject = getRelatedObject()
+  const uploadedListRef = useRef(uploadedList)
+
+  useEffect(() => {
+    if (uploadedListRef.current !== uploadedList) {
+      uploadedListRef.current = uploadedList
+    } else {
+      if (uploadedList.length > 0 && saveRelatedObject()) {
```
Comment (midmarch) [inline client/src/components/Attachment/UploadAttachment.js:44]: Using an `&&` with an unresolved promise?

File: client/src/components/Attachment/UploadAttachment.js [hunk 17]
```diff
@@ -25,11 +25,29 @@ const GQL_UPDATE_ATTACHMENT = gql`
...
+        for (let i = 0; i < uploadedList.length; i++) {
+          save(uploadedList[i], saveRelatedObject(), true)
```
Comment (midmarch) [inline client/src/components/Attachment/UploadAttachment.js:46]: `saveRelatedObject()` called in a loop?

File: anet-dictionary.yml [hunk 18]
```diff
@@ -327,7 +327,6 @@ fields:
     mimeTypes:
       - application/pdf
       - image/jpeg
-      - image/jpg
       - image/png
       - text/plain
```
Comment (midmarch) [inline anet-dictionary.yml:331]: I though we also would add support for webp
```suggestion
      - image/webp
      - text/plain
```

**Review by midmarch — CHANGES_REQUESTED** [review]
Uploading an attchment did not work, see comment regading handling upload error.

File: client/src/components/Attachment/UploadAttachment.js (hunk 13 above)
Comment (mk-simsoft) [reply client/src/components/Attachment/UploadAttachment.js:141]: This error happened cause, there is no control for null default classifiation for updateAttachment query so I added control its okay now.

**Review by hyesilAnka — CHANGES_REQUESTED** [review]
Please handle following scenario:
- Add new attachment the location
- Delete this new attachment
- Then try to add same attachement again. it fails, please solve it.

File: client/src/components/Attachment/UploadAttachment.js [hunk 19]
```diff
@@ -25,26 +25,56 @@ const GQL_UPDATE_ATTACHMENT = gql`
...
+    // if base entity is not created, the attachment can not be associated
+    if (!relatedObjectUuid) {
+      return
+    }
+    // Update only object not associated with the base entity, as in report
+    const updateList = uploadedList.filter(
+      list => !list.attachmentRelatedObjects[0].relatedObjectUuid
+    )
+    for (let i = 0; i < updateList.length; i++) {
+      save(updateList[i], relatedObjectUuid, true)
```
Comment (hyesilAnka) [inline client/src/components/Attachment/UploadAttachment.js:60]: You could think in case of save is not succesfull. Like in attachment save.

**Review by hyesilAnka — CHANGES_REQUESTED** [review]
Please handle following scenario:
- Add new attachment
- Then open attachment dialog and click another folder 
- Then click cancel button. it will give error:
image

Console output is:
image

File: client/src/pages/attachments/Form.js [hunk 20]
```diff
@@ -247,7 +248,12 @@ const AttachmentForm = ({ edit, title, initialValues }) => {
 
   function save(values, form) {
     const attachment = Attachment.filterClientSideFields(values, "content")
+    const updatedRelatedObjects = relatedObjects.map(attach => ({
+      relatedObjectType: attach.relatedObjectType,
+      relatedObjectUuid: attach.relatedObjectUuid
+    }))
     attachment.classification = values.classification
+    attachment.attachmentRelatedObjects = updatedRelatedObjects
```
Comment (gjvoosten) [inline client/src/pages/attachments/Form.js:256]: The form parameter is unused; instead I suggest passing the related objects as a parameter. Attachment never contains a "content" field. Also the reassignment of `attachment.classification` can be removed. Code can also be simplified a bit by using destructuring, and removing the unnecessary `const` by assigning it directly to the attachment:
```
  function save(values, attachmentRelatedObjects) {
    const attachment = Attachment.filterClientSideFields(values)
    attachment.attachmentRelatedObjects = attachmentRelatedObjects.map(
      ({ relatedObjectType, relatedObjectUuid }) => ({
        relatedObjectType,
        relatedObjectUuid
      })
    )
```

File: client/src/components/Attachment/UploadAttachment.js [hunk 21]
```diff
@@ -36,6 +35,25 @@ const UploadAttachment = ({
...
+    if (edit) {
+      return
+    }
+    // if base entity is not created, the attachment can not be associated
+    if (!relatedObjectUuid) {
+      return
+    }
+    // Update only object not associated with the base entity, as in report
+    const updateList = uploadedList.filter(
+      list => !list.attachmentRelatedObjects[0].relatedObjectUuid
```
Comment (gjvoosten) [inline client/src/components/Attachment/UploadAttachment.js:50]: It is unclear to me what this effect is supposed to accomplish. And this specific statement implies that attachments can have only one related object, which is not the general case.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/Attachment/UploadAttachment.js (hunk 19 above)
Comment (midmarch) [reply client/src/components/Attachment/UploadAttachment.js:53]: Yes error needs to be handles, 

and the starting point is to stop ignoring the returned promise by `save()`  If you save parallel, you may change the `for` loop to `.map()` and handle the result with `Promise.all()`. 

Ref: https://www.techiediaries.com/promise-all-map-async-await-example/

Or turn into serial saving by adding await in front of `save`.

**Review by midmarch — CHANGES_REQUESTED** [review]
Basic behaviour uploading attachement looks good now.

Please check how promises are handled while calling `async save()`

File: client/src/pages/locations/Form.js [hunk 22]
```diff
@@ -218,28 +205,23 @@ const LocationForm = ({ edit, title, initialValues, notesComponent }) => {
...
+                        relatedObjectUuid={values.uuid}
+                      />
                     }
-                  }}
-                  onHandleBlur={() => {
-                    // validation will be done by setFieldValue
-                    setFieldTouched("description", true, false)
-                  }}
-                  widget={<RichTextEditor className="description" />}
-                />
```
Comment (hyesilAnka) [inline client/src/pages/locations/Form.js:242]: Please, take this changes back

**Review by hyesilAnka — CHANGES_REQUESTED** [review]

File: client/tests/webdriver/baseSpecs/showLocation.spec.js [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+import ShowLocation from "../pages/location/showLocation.page"
+
+const LOCATION_UUID = "e5b3a4b9-acf7-4c79-8224-f248b9a7215d"
+
+describe("Show location page", () => {
+  beforeEach("Open the show location page", async() => {
+    await ShowLocation.open(LOCATION_UUID)
+  })
+
+  describe("WHen on the show page of a location with attachment(s)", () => {
```
Comment (hyesilAnka) [inline client/tests/webdriver/baseSpecs/showLocation.spec.js:11]: typo

**Review by hyesilAnka — CHANGES_REQUESTED** [review]

File: client/tests/webdriver/baseSpecs/createNewLocation.spec.js [hunk 24]
```diff
@@ -2,7 +2,6 @@ import { expect } from "chai"
 import CreateNewLocation from "../pages/location/createNewLocation.page"
 import {
   BAD_LAT_LNG_VAL,
-  DESCRIPTION,
```
Comment (hyesilAnka) [inline client/tests/webdriver/baseSpecs/createNewLocation.spec.js:5]: Please, revert this changes

File: client/tests/webdriver/pages/location/createNewLocation.page.js [hunk 25]
```diff
@@ -116,12 +112,6 @@ class CreateNewLocation extends Page {
     await (await this.getForm()).waitForExist()
     await (await this.getForm()).waitForDisplayed()
   }
-
-  async fillLocationDescription(description) {
-    await (await this.getDescriptionInput()).click()
-    await browser.keys(description)
-    await browser.pause(300)
-  }
```
Comment (hyesilAnka) [inline client/tests/webdriver/pages/location/createNewLocation.page.js:124]: please, revert this changes

File: client/tests/webdriver/pages/location/createNewLocation.page.js [hunk 26]
```diff
@@ -15,10 +15,6 @@ class CreateNewLocation extends Page {
     return browser.$('//div[contains(text(),"name is a required field")]')
   }
 
-  async getDescriptionInput() {
-    return browser.$("#fg-description .editable")
-  }
```
Comment (hyesilAnka) [inline client/tests/webdriver/pages/location/createNewLocation.page.js:20]: please, revert this changes

**Review by hyesilAnka — CHANGES_REQUESTED** [review]

**Review by hyesilAnka — APPROVED** [review]

File: client/src/components/Attachment/Attachment.css [hunk 27]
```diff
@@ -94,17 +94,91 @@
...
+
+  .image-preview {
+    background-size: cover;
+    background-repeat: no-repeat;
+    height: 200px; /* Ayarlamaları gerektiğiniz yükseklik */
+  }
+
+  .info-line {
+    font-size: 12px;
+  }
```
Comment (gjvoosten) [inline client/src/components/Attachment/Attachment.css:146]: I would remove this and leave the font-size alone.

File: client/src/components/Attachment/Attachment.css [hunk 28]
```diff
@@ -94,17 +94,91 @@
   border-radius: 0.10rem 0.10rem 0 0;
 }
 
+#attachmentCardList {
```
Comment (gjvoosten) [inline client/src/components/Attachment/Attachment.css:97]: I would change this from an id to a className.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/Attachment/Attachment.css [hunk 29]
```diff
@@ -94,17 +94,91 @@
...
+  }
+}
+/* For bigger devices than tablets */
+@media (min-width: 769px) and (max-width: 1260px) {
+  .attachmentCard {
+    flex: 0 0 calc(50%);
+    max-height: 155px;
+  }
+
+  .image-preview {
```
Comment (gjvoosten) [inline client/src/components/Attachment/Attachment.css:165]: So now we have classes `image-preview` and `img-preview`; why?

**Review by gjvoosten — APPROVED** [review]
